Userland/Libraries/LibWeb/Painting/InlinePaintable.cpp

```diff
@@ -203,10 +203,50 @@ TraversalDecision InlinePaintable::hit_test(CSSPixelPoint position, HitTestType
         if (fragment_absolute_rect.contains(position_adjusted_by_scroll_offset)) {
             if (fragment.paintable().hit_test(position, type, callback) == TraversalDecision::Break)
                 return TraversalDecision::Break;
-            auto hit_test_result = HitTestResult { const_cast<Paintable&>(fragment.paintable()), fragment.text_index_at(position_adjusted_by_scroll_offset.x()) };
+            auto hit_test_result = HitTestResult { const_cast<Paintable&>(fragment.paintable()), fragment.text_index_at(position_adjusted_by_scroll_offset.x()), 0, 0 };
             if (callback(hit_test_result) == TraversalDecision::Break)
                 return TraversalDecision::Break;
         }
+
+        // If we reached this point, the position is not within the fragment. However, the fragment start or end might be the place to place the cursor.
+        // This determines whether the fragment is a good candidate for the position. The last such good fragment is chosen.
+        // The best candidate is either the end of the line above, the beginning of the line below, or the beginning or end of the current line.
+        // We arbitrarily choose to consider the end of the line above and ignore the beginning of the line below.
+        // If we knew the direction of selection, we could make a better choice.
+        auto const hit_result_outside_fragment = [fragment_absolute_rect, position_adjusted_by_scroll_offset, fragment]() -> Optional<HitTestResult> {
+            if (fragment_absolute_rect.bottom() - 1 <= position_adjusted_by_scroll_offset.y()) { // fully below the fragment
+                return HitTestResult {
```

Comment:
What should happen in the case that the point is below ***and*** to the left or right of the fragment. Do we need to calculate both the horizontal and vertical distances before returning a result, or is it better to discount these points entirely?

Would something like this work:

```c++
if (!fragment_absolute_rect.contains(position_adjusted_by_scroll_offset)) {
    CSSPixels vertical_distance = 0;
    CSSPixels horizontal_distance = 0;
    if (fragment_absolute_rect.left() > position_adjusted_by_scroll_offset.x())
        horizontal_distance = fragment_absolute_rect.left() - position_adjusted_by_scroll_offset.x();
    else if (fragment_absolute_rect.right() < position_adjusted_by_scroll_offset.x())
        horizontal_distance = position_adjusted_by_scroll_offset.x() - fragment_absolute_rect.right();

    if (fragment_absolute_rect.top() > position_adjusted_by_scroll_offset.y())
        vertical_distance = fragment_absolute_rect.top() - position_adjusted_by_scroll_offset.y();
    else if (fragment_absolute_rect.bottom() < position_adjusted_by_scroll_offset.y())
        vertical_distance = position_adjusted_by_scroll_offset.y() - fragment_absolute_rect.bottom();

    return HitTestResult {
	    .paintable = const_cast<Paintable &>(fragment.paintable()),
    	.index_in_node = fragment.start() + fragment.length(),
	    .vertical_distance = vertical_distance,
    	.horizontal_distance = horizontal_distance,
    };
}
```

This is more of a question than a suggestion from me, I'm not sure if you'd get better results doing this or not.

Reply:
Thanks for reviewing. There is a reason that horizontal and vertical distances are calculated secretly, but I can see how the code doesn’t make that clear. I’ve attempted to make this clearer, but I couldn’t really think of a super nice way of doing that, so I’ve mostly just resorted to explaining it in a comment. If you can think of a better way of factoring this, I’d be happy to hear it.

Basically the problem is that the closest fragment is not always the one you’d expect your cursor to be placed at. Here’s two examples of that:

![horizontal-first](https://github.com/user-attachments/assets/64002d34-8845-405e-94a2-8720b4c997f9)
![only-under](https://github.com/user-attachments/assets/8267404d-ea11-4ccb-b12e-ef459d72ef7e)


So to solve this, I came up with the rule that, if your cursor is on the same line as a fragment, that will always be picked. And it will always pick a fragment above the cursor if there is one.